share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml

```diff
@@ -148,7 +148,7 @@ spack:
   - nco
   - netlib-scalapack
   - omega-h
-  - openmpi
+  # - openmpi
```

Comment:
Does reformulating `conflicts()` as I suggested (combining with `vendors()`) remove the need for this?

Reply:
No. This is related to the fact that:
```yaml
spack:
  specs:
  - openmpi
  packages:
    mpi:
      require: mpich
```
is unsat in this PR, while it can be concretized on `develop`. It seems like a bug on `develop` to me.

Reply:
I don't think so -- the way I read that is "if the user or a package requested `mpi`, make it `mpich`".  But the user didn't request `mpi` here.  They requested `openmpi`, so there should be no issue with the provider requirement (`openmpi` doesn't require `mpi`, even if it is one).  I think it's important for the virtual constraints to *only* trigger when packages are being used *to satisfy a virtual*.

Reply:
I can double check this (untested) behavior on `develop` and try replicate it 1:1. Note though that:
```yaml
spack:
  specs:
  - openmpi
  - mpich
  - hdf5
  packages:
    mpi:
      require: mpich
  concretizer:
    unify: true
```
is unsat in this PR, while on `develop` it flips the default value of the `mpi` variant for `hdf5` - so no `mpi` is used.[^1] I think the behavior in this PR is maybe less unexpected?

[^1]: `mpich` and `openmpi` can be concretized, but then no other spec can use `mpi` without triggering a conflict between the two providers.

Reply:
Just my two cents: my mental model for `unify: true` is an anonymous package with the `specs` list as direct deps. I wouldn't expect `specs: [mpich, openmpi]` to concretize at all with unify:true, and instead follow the 1 virtual per DAG rule.

Reply:
@tgamblin See https://github.com/spack/spack/pull/38447#issuecomment-1679064680 for the fix to this comment. I still think the required behavior might be unexpected, but we can fine tune that when we'll review #35322 (which I plan to do once this is merged).

Reply:
I missed @haampie's comment above but I'm not opposed to changing this semantic.  I do think it should be in a follow-on PR though, and easier to surface than a discussion inside this concretizer rework.

My worry is that there are actually a lot of things that *happen* to provide various interfaces, and there shouldn't necessarily be a conflict between them if they're requested independently.  My definition for `unify: true` is one node per package and one provider *within* a process, and there's no reason with `RPATH` semantics for roots that aren't providing `mpi` to interfere with `mpi` being used elsewhere.  But let's hash that out separately -- I think it's more important here to not break existing stacks, and we can be more explicit about potentially breaking concretizer changes with a smaller PR.